modules/contact/src/constraints/RANFSNormalMechanicalContact.C

```diff
@@ -0,0 +1,415 @@
+        }
+
+        _node_to_contact_lm.insert(std::make_pair(_current_node->id(), _res_vec * _pinfo->_normal));
+        _node_to_tied_lm.insert(std::make_pair(_current_node->id(), _res_vec(_component)));
+      }
+      else
+      {
+        // We need the matrix to be assembled so we get the correct Jacobian entries
+        if (!_jacobian->closed())
+          _jacobian->close();
+
+        std::vector<dof_id_type> cols;
+        std::vector<Number> values;
+
+        for (auto & d_to_v : _dof_number_to_value)
+          d_to_v.clear();
+
+        mooseAssert(_vars.size() == _dof_number_to_value.size() &&
+                        _vars.size() == _var_objects.size(),
+                    "Somehow the sizes of our variable containers got out of sync");
+        for (MooseIndex(_var_objects) i = 0; i < _var_objects.size(); ++i)
+        {
+          auto slave_dof_number = _current_node->dof_number(0, _vars[i], 0);
+
+          _jacobian->get_row(slave_dof_number, cols, values);
+          mooseAssert(cols.size() == values.size(),
+                      "The size of the dof container and value container are different");
+
+          for (MooseIndex(cols) j = 0; j < cols.size(); ++j)
+            _dof_number_to_value[i].insert(
+                std::make_pair(cols[j], values[j] / _var_objects[i]->scalingFactor()));
+        }
+      }
+
+      mooseAssert(_node_to_contact_lm.find(_current_node->id()) != _node_to_contact_lm.end(),
+                  "The node " << _current_node->id()
+                              << " should map to a contact lagrange multiplier");
+      mooseAssert(_node_to_tied_lm.find(_current_node->id()) != _node_to_tied_lm.end(),
+                  "The node " << _current_node->id()
+                              << " should map to a tied lagrange multiplier");
+      _contact_lm = _node_to_contact_lm[_current_node->id()];
+      _tied_lm = _node_to_tied_lm[_current_node->id()];
+
+      // Check to see whether we've locked a ping-ponging node
+      if (_ping_pong_slave_node_to_master_node.find(_current_node->id()) ==
+          _ping_pong_slave_node_to_master_node.end())
+      {
+        if (_contact_lm > -_pinfo->_distance)
+        {
+          // Ok, our math is telling us we should apply the constraint, but what if we are
+          // ping-ponging back and forth between different master faces?
+
+          // This only works for a basic line search! Write assertion here
+          if (_subproblem.computingNonlinearResid())
+          {
+            auto & master_elem_sequence = _node_to_master_elem_sequence[_current_node->id()];
+            mooseAssert(
+                _current_master == _pinfo->_elem,
+                "The current master element and the PenetrationInfo object's element should "
+                "be the same");
+            master_elem_sequence.push_back(_pinfo->_elem);
+
+            if (master_elem_sequence.size() >= 5 &&
```

Comment:
It's definitely a heuristic. In my testing experience, if a node ping-pongs back and forth 5 times then it will ping pong indefinitely. However, if it goes 3 times for example, it is not guaranteed to ping-pong indefinitely and the Newton iteration may naturally resolve the correct face dofs that need to be involved in the constraint.

Reply:
It will be helpful to put these comments into the code

Reply:
👍